### PR TITLE
use rapidfuzz instead of fuzzywuzzy

### PR DESCRIPTION
**Pull request by maxbachmann**
FuzzyWuzzy is GPLv2 licensed which would force you to licence the whole project under GPLv2.
For this reason this Pullrequest replaces FuzzyWuzzy with  [rapidfuzz](https://github.com/rhasspy/rapidfuzz) which is implementing the same algorithm but is based on a version of fuzzywuzzy that was MIT Licensed.
Rapidfuzz is:
- Mit licensed so it can be used with the license used by this project
- Is faster than FuzzyWuzzy

### CODE REVIEW

File: grayskull/license/discovery.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def match_license(name: str) -> dict:
     all_licenses = get_all_licenses_from_spdx()
     name = re.sub(r"\s+license\s*", "", name.strip(), flags=re.IGNORECASE)
 
-    best_matches = process.extractBests(name, _get_all_license_choice(all_licenses))
```
Comment (maxbachmann) [inline grayskull/license/discovery.py]: extract and extractBests in FuzzyWuzzy are the same with the difference that extractBests allows the use of score_cutoff which is not possible with extract.
RapidFuzz is based on an older version of FuzzyWuzzy that did not have extractBests, but has the score_cutoff argument for extract. So extract does the same extractBests does

**Comment by marcelotrevisani** [comment]
Hi @maxbachmann ,
Thanks for letting me know that! I think we don't have this package on ``conda-forge``, I will add this package there. After that we can check if the tests are passing, if all the tests passes I think we can go for it!

**Comment by marcelotrevisani** [comment]
Could you please add the ``sdist`` on PyPI for ``rapidfuzz``?
https://github.com/rhasspy/rapidfuzz/issues/12

**Comment by maxbachmann** [comment]
Should be added. I started working on conda-forge: https://github.com/conda-forge/staged-recipes/pull/11111#issuecomment-603177463 but I never did before and so have no clue what I screwed up. So I would appreciate your help on this.

**Comment by maxbachmann** [comment]
It's on conda forge now https://anaconda.org/conda-forge/rapidfuzz

**Comment by marcelotrevisani** [comment]
Just closed and  reopened to trigger the CI

**Comment by maxbachmann** [comment]
Oh a segfault in rapidfuzz sounds bad ...
Do you know how extract is called in this case so I can reproduce it?

**Comment by maxbachmann** [comment]
Hm it should be fixed in the latest version on conda-forge, but apparently it still selects the old version

**Comment by maxbachmann** [comment]
Hm still using the old version. It should work when it picks the newer package from conda-forge again ;)

**Comment by maxbachmann** [comment]
I will add some unit tests to rapidfuzz and then try this again ^^

**Comment by maxbachmann** [comment]
Hm one of the errors is a bit of luck based behaviour in your code:
You test
```
 assert get_short_license_id("GPL 2.0") == "GPL-2.0-or-later"
```
this happens in two matching steps first it get's 5 elements using extract
```
[('GPL-2.0-only', 90), ('GPL-2.0-or-later', 90), ('LGPL-2.0-only', 90), ('LGPL-2.0-or-later', 90), ('Academic Free License v1.2', 86)]
```
I however get the result
```
[('GPL-2.0-only', 90.0), ('GPL-2.0-or-later', 90.0), ('LGPL-2.0-only', 90.0), ('LGPL-2.0-or-later', 90.0), ('EPL-2.0', 85.71428571428572)]```
which is actually the correct one, since Academic Free License v1.2' has only about 85.5 and just gets used by fuzzywuzzy since they are both at 86% -> it is a bit of luck which one you get

afterwards it matches using token_sort_ratio which results in
```
>>> fuzz.token_sort_ratio("GPL-2.0-or-later", "GPL 2.0")
61
>>> fuzz.token_sort_ratio("EPL-2.0", "GPL 2.0")
86
```
thats why rapidfuzz returns the EPL-2.0. with FuzzyWuzzy you appear to have luck, since the sorting algorithm appears to place the 'Academic Free License v1.2' earlier in the list. However thats most likely not guaranteed and could therefor change at any time.

**Comment by marcelotrevisani** [comment]
> Hm one of the errors is a bit of luck based behaviour in your code:
> You test
> 
> ```python
>  assert get_short_license_id("GPL 2.0") == "GPL-2.0-or-later"
> ```
> 
> this happens in two matching steps first it get's 5 elements using extract
> 
> ```
> [('GPL-2.0-only', 90), ('GPL-2.0-or-later', 90), ('LGPL-2.0-only', 90), ('LGPL-2.0-or-later', 90), ('Academic Free License v1.2', 86)]
> ```
> 
> I however get the result
> 
> ```
> [('GPL-2.0-only', 90.0), ('GPL-2.0-or-later', 90.0), ('LGPL-2.0-only', 90.0), ('LGPL-2.0-or-later', 90.0), ('EPL-2.0', 85.71428571428572)]
> ```
> 
> which is actually the correct one, since Academic Free License v1.2' has only about 85.5 and just gets used by fuzzywuzzy since they are both at 86% -> it is a bit of luck which one you get
> 
> afterwards it matches using token_sort_ratio which results in
> 
> ```
> > fuzz.token_sort_ratio("GPL-2.0-or-later", "GPL 2.0")
> 61
> > fuzz.token_sort_ratio("EPL-2.0", "GPL 2.0")
> 86
> ```
> 
> thats why rapidfuzz returns the EPL-2.0. with FuzzyWuzzy you appear to have luck, since the sorting algorithm appears to place the 'Academic Free License v1.2' earlier in the list. However thats most likely not guaranteed and could therefor change at any time.

Yeah, that seems to be a problem indeed, what kind of matching scorer do you recommend instead?

**Comment by maxbachmann** [comment]
I think partial_ratio would be better suited, since it searchs for the best matching substrings. So e.g. with the GPL 2.0 example all results that have the word gpl 2.0 in them have a score of 100.
 Then afterwards a second ratio to find the correct one of them which could be done with a token_set_ratio.
```console
> process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.ratio)
[('GPL-2.0-only', 73.6842105263158), ('LGPL-2.0-only', 70.0), ('GPL-2.0-or-later', 60.86956521739131), ('LGPL-2.0-or-later', 58.33333333333333)]
> process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.token_set_ratio)
[('GPL-2.0-only', 100.0), ('GPL-2.0-or-later', 100.0), ('LGPL-2.0-only', 70.0), ('LGPL-2.0-or-later', 60.0)]
> process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.token_sort_ratio)
[('GPL-2.0-only', 73.6842105263158), ('LGPL-2.0-only', 70.0), ('GPL-2.0-or-later', 60.86956521739131), ('LGPL-2.0-or-later', 58.33333333333333)]
> process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.partial_ratio)
[('GPL-2.0-only', 100.0), ('GPL-2.0-or-later', 100.0), ('LGPL-2.0-only', 100.0), ('LGPL-2.0-or-later', 100.0)]
```

**Comment by marcelotrevisani** [comment]
> I think partial_ratio would be better suited, since it searchs for the best matching substrings. So e.g. with the GPL 2.0 example all results that have the word gpl 2.0 in them have a score of 100.
> Then afterwards a second ratio to find the correct one of them which could be done with a token_set_ratio.
> 
> ```
> > process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.ratio)
> [('GPL-2.0-only', 73.6842105263158), ('LGPL-2.0-only', 70.0), ('GPL-2.0-or-later', 60.86956521739131), ('LGPL-2.0-or-later', 58.33333333333333)]
> > process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.token_set_ratio)
> [('GPL-2.0-only', 100.0), ('GPL-2.0-or-later', 100.0), ('LGPL-2.0-only', 70.0), ('LGPL-2.0-or-later', 60.0)]
> > process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.token_sort_ratio)
> [('GPL-2.0-only', 73.6842105263158), ('LGPL-2.0-only', 70.0), ('GPL-2.0-or-later', 60.86956521739131), ('LGPL-2.0-or-later', 58.33333333333333)]
> > process.extract("gpl 2.0", ['GPL-2.0-only', 'GPL-2.0-or-later', 'LGPL-2.0-only', 'LGPL-2.0-or-later'], scorer=fuzz.partial_ratio)
> [('GPL-2.0-only', 100.0), ('GPL-2.0-or-later', 100.0), ('LGPL-2.0-only', 100.0), ('LGPL-2.0-or-later', 100.0)]
> ```

I do not remember why I wasn't using that.... but if it works fine that is good for me :D
Are you wiling to modify that as well? or should I fix it first?

**Comment by maxbachmann** [comment]
I can update this aswell

File: grayskull/license/discovery.py [hunk 2]
```diff
@@ -61,14 +61,20 @@ def match_license(name: str) -> dict:
...
         best_matches = [l[0] for l in best_matches if not l[0].endswith("-only")]
+
         if best_matches:
-            spdx_license = process.extractOne(
-                name, best_matches, scorer=token_sort_ratio
-            )
+            best_matches = process.extract(name, best_matches, scorer=token_set_ratio)
+            spdx_license = best_matches[0]
+            best_matches = [l[0] for l in best_matches if l[1] >= spdx_license[1]]
+            if len(best_matches) > 1:
```
Comment (marcelotrevisani) [inline grayskull/license/discovery.py:73]: nice! thanks for changing this as well! :)

**Comment by marcelotrevisani** [comment]
I think that is good to go!
Sorry for my delay, I have been a bit busy these days :/

Thanks a lot for your contribution, I will generate a new version in a couple of minutes with this PR

**Pull request merged**